feat: use undici as fetch polyfill

jacob-ebey opened the pull request:
Use undici as our fetch polyfill going forward. This will allow us to firstly (and my my favorite part), stop maintaining our own, and secondly and arguably more important, will allow us to smoothly upgrade our users to the Node.js fetch when they are on a node 20+ as they will already be using it via the polyfill. Undici is the node fetch implementation, and now all our installGlobals() will be doing is giving you the latest version regardless of your node version, and I think that's pretty sweet.

<!--

👋 Hey, thanks for your interest in contributing to Remix!

If this is a simple docs change, go ahead and delete all this.

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a
lot of time and effort into a new feature. To avoid this from happening, we
request that contributors create a
[Feature Request discussion](https://github.com/remix-run/remix/discussions/new?category=ideas)
to first discuss any significant new features.

https://github.com/remix-run/remix/blob/main/CONTRIBUTING.md

Please fill in or delete each item below:

-->

Closes: #

- [ ] Docs
- [ ] Tests

Testing Strategy:

<!--
Please explain how you tested this. For example:

> This test covers this code: <link to test>

Or

> I opened up my windows machine and ran this script:
>
> ```
> npx create-remix@0.0.0-experimental-7e420ee3 my-test
> cd my-test
> npm run dev
> ```
-->

mjackson commented on packages/remix-react/__tests__/setup.ts:
```diff
@@ -1,3 +1,8 @@
+global.WritableStream = require("stream/web").WritableStream;
 
-installGlobals();
+require("@remix-run/node").installGlobals();
```
It seems like `installGlobals` should also install `TextDecoder`, `TextEncoder`, `ReadableStream`, and `WritableStream`, no?

jacob-ebey replied:
This is a JSDom issue here.

mjackson reviewed (commented):
this looks pretty sweet! 🙌  Nice work!

One small question: It seems like in addition to installing the `fetch` family in our polyfill, we should also install `ReadableStream` and `WritableStream`, as well as any other stream classes provided by undici. Do you agree @jacob-ebey ?

garand commented:
Keep an eye out for this bug. I moved to Undici on a project and had to switch off because the API I was working with wasn't including a content-length.

https://github.com/nodejs/undici/pull/1540

jacob-ebey commented:
> Keep an eye out for this bug. I moved to Undici on a project and had to switch off because the API I was working with wasn't including a content-length.
> 
> [nodejs/undici#1540](https://github.com/nodejs/undici/pull/1540)

I added a test against this and it appears to have been fixed at some point: https://github.com/nodejs/undici/pull/2995
